core/services/log/orm.go

```diff
@@ -81,17 +129,118 @@ func (o *orm) MarkBroadcastConsumed(blockHash common.Hash, blockNumber uint64, l
+	if err != nil {
+		return nil, err
+	}
+	minPending, err := o.GetBroadcastsPending(qopts...)
+	if err != nil {
+		return nil, err
+	}
+	if minUnconsumed == nil {
+		return minPending, nil
+	}
+	if minPending == nil || *minUnconsumed < *minPending {
+		minPending = minUnconsumed
+		if err := o.SetBroadcastsPending(minPending, qopts...); err != nil {
+			return nil, err
+		}
+	}
+	if err := o.removeUnconsumed(qopts...); err != nil {
+		return nil, err
+	}
+	return minPending, nil
+}
+
+func (o *orm) SetBroadcastsPending(blockNumber *int64, qopts ...postgres.QOpt) error {
+	q := postgres.NewQ(o.db, qopts...)
+	res, err := q.Exec(`
+        INSERT INTO log_broadcasts_pending (evm_chain_id, block_number, created_at, updated_at) VALUES ($1, $2, NOW(), NOW())
+		ON CONFLICT (evm_chain_id) DO UPDATE SET block_number = $3, updated_at = NOW() 
+    `, o.evmChainID, null.IntFromPtr(blockNumber), null.IntFromPtr(blockNumber))
+	if err != nil {
+		return errors.Wrap(err, "while setting pending broadcast block number")
+	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return errors.Wrap(err, "SetBroadcastsPending failed")
+	}
+	if rowsAffected == 0 {
+		return errors.Wrap(err, "while setting pending broadcast block number")
+	}
+	return nil
+}
+
+func (o *orm) GetBroadcastsPending(qopts ...postgres.QOpt) (*int64, error) {
+	q := postgres.NewQ(o.db, qopts...)
+	var blockNumber null.Int
+	err := q.QueryRowx(`
+        SELECT block_number FROM log_broadcasts_pending WHERE evm_chain_id = $1
+    `, o.evmChainID).Scan(&blockNumber)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, nil
+	} else if err != nil {
+		return nil, errors.Wrap(err, "failed to get broadcasts pending number")
+	}
+	if blockNumber.Valid {
+		return &blockNumber.Int64, nil
+	}
+	return nil, nil
+}
+
+func (o *orm) getUnconsumedMinBlock(qopts ...postgres.QOpt) (*int64, error) {
+	q := postgres.NewQ(o.db, qopts...)
+	var blockNumber null.Int
+	err := q.QueryRowx(`
+        SELECT min(block_number) FROM log_broadcasts
+			WHERE evm_chain_id = $1
+			AND consumed = false
+			AND block_number IS NOT NULL
+    `, o.evmChainID).Scan(&blockNumber)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, nil
+	} else if err != nil {
+		return nil, errors.Wrap(err, "failed to get unconsumed broadcasts min block number")
+	}
+	if blockNumber.Valid {
+		return &blockNumber.Int64, nil
+	}
+	return nil, nil
+}
+
+func (o *orm) removeUnconsumed(qopts ...postgres.QOpt) error {
+	q := postgres.NewQ(o.db, qopts...)
+	res, err := q.Exec(`
+        DELETE FROM log_broadcasts
+			WHERE evm_chain_id = $1
+			AND consumed = false
+			AND block_number IS NOT NULL
+    `, o.evmChainID)
+	if err != nil {
+		return errors.Wrap(err, "while deleting unconsumed broadcasts")
+	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return errors.Wrap(err, "while deleting unconsumed broadcasts")
+	}
+	if rowsAffected == 0 {
+		return errors.Wrap(err, "while deleting unconsumed broadcasts")
```

Comment:
i guess these 3 error messages could be slightly different

Reply:
Yeah I will revisit the error handling :+1: